diff --git a/whisper.cpp/ggml/src/ggml-cpu/CMakeLists.txt b/whisper.cpp/ggml/src/ggml-cpu/CMakeLists.txt
index e52e050a..c1000c16 100644
--- a/whisper.cpp/ggml/src/ggml-cpu/CMakeLists.txt
+++ b/whisper.cpp/ggml/src/ggml-cpu/CMakeLists.txt
@@ -106,7 +106,7 @@ function(ggml_add_cpu_backend_variant_impl tag_name)
             )
 
         if (MSVC AND NOT CMAKE_C_COMPILER_ID STREQUAL "Clang")
-            message(FATAL_ERROR "MSVC is not supported for ARM, use clang")
+            list(APPEND ARCH_FLAGS /arch:armv8.7)
         else()
             check_cxx_compiler_flag(-mfp16-format=ieee GGML_COMPILER_SUPPORTS_FP16_FORMAT_I3E)
             if (NOT "${GGML_COMPILER_SUPPORTS_FP16_FORMAT_I3E}" STREQUAL "")
diff --git a/whisper.cpp/ggml/src/ggml-hexagon/ggml-hexagon.cpp b/whisper.cpp/ggml/src/ggml-hexagon/ggml-hexagon.cpp
index cabd301a..22361297 100644
--- a/whisper.cpp/ggml/src/ggml-hexagon/ggml-hexagon.cpp
+++ b/whisper.cpp/ggml/src/ggml-hexagon/ggml-hexagon.cpp
@@ -8,6 +8,7 @@
 #include <atomic>
 #include <chrono>
 #include <mutex>
+#include <stdexcept>
 #include <string>
 
 #ifdef _WIN32
@@ -3481,11 +3482,25 @@ static const char * ggml_backend_hexagon_device_get_description(ggml_backend_dev
     GGML_UNUSED(dev);
 }
 
+// ~2GB per session for now
+#define GGML_HEXAGON_SESSION_MEMORY_DEFAULT (2ULL * 1024 * 1024 * 1024)
+// Max to 3.5GB
+#define GGML_HEXAGON_SESSION_MEMORY_MAX (3ULL * 1024 * 1024 * 1024  +  512ULL * 1024 * 1024)
+
 static void ggml_backend_hexagon_device_get_memory(ggml_backend_dev_t dev, size_t * free, size_t * total) {
-    // ~2GB per session for now
-    *free  = 2ULL * 1024 * 1024 * 1024;
-    *total = *free;
+    const char * str_mem = getenv("GGML_HEXAGON_SESSION_MEMORY");
+    if (str_mem) {
+        *free = std::stoull(str_mem);
+        if (*free < GGML_HEXAGON_SESSION_MEMORY_DEFAULT) {
+            *free = GGML_HEXAGON_SESSION_MEMORY_DEFAULT;
+        } else if (*free > GGML_HEXAGON_SESSION_MEMORY_MAX) {
+            *free = GGML_HEXAGON_SESSION_MEMORY_MAX;
+        }
+    } else {
+        *free = GGML_HEXAGON_SESSION_MEMORY_DEFAULT;
+    }
 
+    *total = *free;
     GGML_UNUSED(dev);
 }
 
@@ -3666,10 +3681,17 @@ ggml_hexagon_registry::ggml_hexagon_registry(ggml_backend_reg_t reg) {
         }
     }
 
+#if defined(__ANDROID__)
     if(opt_arch < 75) {
         opt_ndev = 1;
-        GGML_LOG_WARN("ggml-hex: forcing ndev to 1 for SoCs archs lower than v75.\n");
+        GGML_LOG_WARN("ggml-hex: forcing ndev to 1 for SoCs archs lower than v75 for Android.\n");
     }
+#else
+    if(opt_arch < 73) {
+        opt_ndev = 1;
+        GGML_LOG_WARN("ggml-hex: forcing ndev to 1 for SoCs archs lower than v73 for Linux and Windows.\n");
+    }
+#endif
 
     GGML_LOG_INFO("ggml-hex: Hexagon Arch version v%d\n", opt_arch);
 
@@ -3682,6 +3704,8 @@ ggml_hexagon_registry::ggml_hexagon_registry(ggml_backend_reg_t reg) {
         } catch (std::exception const &exc) {
             GGML_LOG_ERROR("ggml-hex: failed to create device/session %zu\n", i);
             devices[i].context = nullptr;
+            opt_ndev = i;
+            break;
         }
     }
 }
diff --git a/whisper.cpp/ggml/src/ggml-vulkan/CMakeLists.txt b/whisper.cpp/ggml/src/ggml-vulkan/CMakeLists.txt
index de01336c..29b1a043 100644
--- a/whisper.cpp/ggml/src/ggml-vulkan/CMakeLists.txt
+++ b/whisper.cpp/ggml/src/ggml-vulkan/CMakeLists.txt
@@ -121,7 +121,7 @@ if (Vulkan_FOUND)
     endif()
 
     # Set up toolchain for host compilation whether cross-compiling or not
-    if (CMAKE_CROSSCOMPILING)
+    if (CMAKE_CROSSCOMPILING OR NOT CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL CMAKE_SYSTEM_PROCESSOR)
         if (GGML_VULKAN_SHADERS_GEN_TOOLCHAIN)
             set(HOST_CMAKE_TOOLCHAIN_FILE ${GGML_VULKAN_SHADERS_GEN_TOOLCHAIN})
         else()
@@ -141,7 +141,7 @@ if (Vulkan_FOUND)
 
     include(ExternalProject)
 
-    if (CMAKE_CROSSCOMPILING)
+    if (CMAKE_CROSSCOMPILING OR NOT CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL CMAKE_SYSTEM_PROCESSOR)
         list(APPEND VULKAN_SHADER_GEN_CMAKE_ARGS -DCMAKE_TOOLCHAIN_FILE=${HOST_CMAKE_TOOLCHAIN_FILE})
         message(STATUS "vulkan-shaders-gen toolchain file: ${HOST_CMAKE_TOOLCHAIN_FILE}")
     endif()
