diff --git a/src/whisper.rn/vendor/whisper.cpp/ggml/src/ggml-vulkan/CMakeLists.txt b/src/whisper.rn/vendor/whisper.cpp/ggml/src/ggml-vulkan/CMakeLists.txt
index d7dec92..69507c5 100644
--- a/src/whisper.rn/vendor/whisper.cpp/ggml/src/ggml-vulkan/CMakeLists.txt
+++ b/src/whisper.rn/vendor/whisper.cpp/ggml/src/ggml-vulkan/CMakeLists.txt
@@ -159,7 +159,7 @@ if (Vulkan_FOUND)
     endif()
 
     # Set up toolchain for host compilation whether cross-compiling or not
-    if (CMAKE_CROSSCOMPILING)
+    if (CMAKE_CROSSCOMPILING OR NOT CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL CMAKE_SYSTEM_PROCESSOR)
         if (GGML_VULKAN_SHADERS_GEN_TOOLCHAIN)
             set(HOST_CMAKE_TOOLCHAIN_FILE ${GGML_VULKAN_SHADERS_GEN_TOOLCHAIN})
         else()
@@ -179,7 +179,7 @@ if (Vulkan_FOUND)
 
     include(ExternalProject)
 
-    if (CMAKE_CROSSCOMPILING)
+    if (CMAKE_CROSSCOMPILING OR NOT CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL CMAKE_SYSTEM_PROCESSOR)
         list(APPEND VULKAN_SHADER_GEN_CMAKE_ARGS -DCMAKE_TOOLCHAIN_FILE=${HOST_CMAKE_TOOLCHAIN_FILE})
         message(STATUS "vulkan-shaders-gen toolchain file: ${HOST_CMAKE_TOOLCHAIN_FILE}")
     endif()
